src/data/equiv/one_sided.lean

```diff
@@ -0,0 +1,215 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+import data.set.function
+
+/-!
+# Functions with a bundled one-sided inverse
+
+This provides `α ↬ β` (`injection`) and `α ↠ β` (`surjection`), for functions associated with a
+specific left- or right-inverse. For the two-sided inverse, use `α ≃ β` (`equiv`).
+
+For cases when you care only that a function is injective, and do not need a specific inverse, use
+`α ↪ β` (`embedding`).
+-/
+
+open function
+
+universes u v w z
+variables {α : Sort u} {β : Sort v} {γ : Sort w} {δ : Sort w}
+
+set_option old_structure_cmd true
+
+/-- `α ↬ β` is the type of functions from `α → β` with an associated left-inverse. -/
+@[nolint has_inhabited_instance]
+structure injection (α : Sort*) (β : Sort*) :=
+(to_fun    : α → β)
+(inv_fun   : β → α)
+(left_inv  : left_inverse inv_fun to_fun)
+
+/-- `α ↠ β` is the type of functions from `α → β` with an associated right-inverse. -/
+@[nolint has_inhabited_instance]
+structure surjection (α : Sort*) (β : Sort*) :=
+(to_fun    : α → β)
+(inv_fun   : β → α)
+(right_inv : right_inverse inv_fun to_fun)
+
+
+infix ` ↬ `:25 := injection
+infix ` ↠ `:25 := surjection
+
+/-! ### Coercion lemmas -/
+
+section coes
+
+namespace injection
+
+instance : has_coe_to_fun (α ↬ β) :=
+⟨_, to_fun⟩
+
+@[simp] lemma to_fun_eq_coe (f : α ↬ β) : f.to_fun = f := rfl
+
+@[simp] lemma coe_mk (f : α → β) (g l) : (injection.mk f g l : α → β) = f :=
+rfl
+
+protected lemma congr_arg {f : α ↬ β} : Π {x x' : α}, x = x' → f x = f x'
+| _ _ rfl := rfl
+
+protected lemma congr_fun {f g : α ↬ β} (h : f = g) (x : α) : f x = g x := h ▸ rfl
+
+end injection
+
+namespace surjection
+
+instance : has_coe_to_fun (α ↠ β) :=
+⟨_, to_fun⟩
+
+@[simp] lemma to_fun_eq_coe (f : α ↠ β) : f.to_fun = f := rfl
+
+@[simp] lemma coe_mk (f : α → β) (g r) : (surjection.mk f g r : α → β) = f :=
+rfl
+
+protected lemma congr_arg {f : α ↠ β} : Π {x x' : α}, x = x' → f x = f x'
+| _ _ rfl := rfl
+
+protected lemma congr_fun {f g : α ↠ β} (h : f = g) (x : α) : f x = g x := h ▸ rfl
+
+end surjection
+
+end coes
+
+/-! ### Symmetry -/
+
+section symm
+
+def injection.symm (f : α ↬ β) : β ↠ α := ⟨f.inv_fun, f.to_fun, f.left_inv.right_inverse⟩
+
+def surjection.symm (f : α ↠ β) : β ↬ α := ⟨f.inv_fun, f.to_fun, f.right_inv.left_inverse⟩
```

Comment:
Should there be a `has_inv` instance for these?

Reply:
No, because `has_inv` is homogenous, and these inverses are heterogenous